common/k8s/src/lease.rs

```diff
@@ -0,0 +1,123 @@
+use chrono::Utc;
+use k8s_openapi::api::coordination::v1::Lease;
+use k8s_openapi::apimachinery::pkg::apis::meta::v1::MicroTime;
+use kube::api::{Api, ListParams, Patch, PatchParams};
+use kube::core::ObjectList;
+use kube::Client;
+use serde::{Deserialize, Serialize};
+
+pub const K8S_STARTUP_LEASE_LABEL: &str = "process=agent-startup";
+pub const K8S_STARTUP_LEASE_RETRY_ATTEMPTS: i32 = 3;
+
+#[derive(Debug, Serialize, Deserialize)]
+struct LeasePatchSpec {
+    spec: LeasePatchValue,
+}
+
+// Non snake case for JSON parsing reasons
+#[allow(non_snake_case)]
+#[derive(Debug, Serialize, Deserialize)]
+struct LeasePatchValue {
+    holderIdentity: Option<String>,
+    acquireTime: MicroTime,
+}
+
+pub async fn get_available_lease(lease_label: &str, lease_client: &Api<Lease>) -> Option<String> {
+    let lease_info = get_lease_list(lease_label, lease_client).await;
+    for lease in lease_info.unwrap().into_iter() {
+        let lease_name = lease.metadata.name.unwrap();
+        match lease.spec.unwrap().holder_identity {
+            Some(lease_owner) => {
+                println!("Lease {} is OWNED by {:?}.", lease_name, lease_owner)
+            }
+            None => {
+                println!("Lease {} NOT OWNED...", lease_name);
+                return Some(lease_name);
+            }
+        }
+    }
+    None
+}
+
+pub async fn claim_lease(
```

Comment:
We could refactor this into a 

```rust
struct Lease {
    lease_name: String,
    ...
}

impl Lease {
    fn acquire(&mut self) -> LeaseGuard {
        // Current contents of claim lease
    }
}

struct LeaseGuard {
    ...
}

impl Drop for LeaseGuard {
    fn drop(self) {
        tokio::spawn(async move {
            // current contents of release_lease
        }
    }
}
```

That way it would be a bit more rustic

Reply:
Thinking we handle this in a follow ticket. Thoughts @c-nixon?